Pod/Classes/LSBinaryCodec.m

```diff
@@ -0,0 +1,324 @@
+                    [LSBinaryCodec readVarintFromProto:protoEnc startingAt:&head];
+                    break;
+                }
+                case WireFormatFixed64:
+                case WireFormatFixed32:
+                case WireFormatLengthDelim:
+                    head += keyInfo.length;
+                default:
+                    if (errorPtr) {
+                        *errorPtr = [NSError errorWithDomain:OTErrorDomain
+                                                        code:OTInvalidCarrierCode
+                                                    userInfo:nil];
+                    }
+            }
+        } else {
+            NSData *msg = [protoEnc subdataWithRange:NSMakeRange(head, keyInfo.length)];
+            LSSpanContext *decoded = dest;
+            bool success = [LSBinaryCodec decodeEmbeddedMessage:msg dest:decoded withError:errorPtr];
+            if (!success) {
+                // TODO: Do something here
+                return NO;
+            }
+            return YES;
+        }
+    }
+
+    if (errorPtr) {
+        *errorPtr = [NSError errorWithDomain:OTErrorDomain code:OTSpanContextCorruptedCode userInfo:nil];
+    }
+    return NO;
+}
+
+#pragma mark - private
+
+typedef struct _keyInfo {
+    Byte fieldType;
+    UInt64 fieldNum;
+    // Non-zero for only fixed-width or length-delimited field types.
+    // 0 for Varint, 8 for fixed64, NumBytes for length-delimited fields.
+    NSUInteger length;
+} KeyInfo;
+
++ (NSData *)encodeInnerMessageForTraceID:(UInt64)traceID
+                                  spanID:(UInt64)spanID
+                                 baggage:(NSDictionary *)baggage {
+    NSMutableData *message = [[NSMutableData alloc] init];
+
+    // fixed64 traceID = 1;
+    [LSBinaryCodec writeTagNumber:1 format:WireFormatFixed64 buffer:message];
+    [LSBinaryCodec writeFixedEncodedUInt64:traceID buffer:message];
+
+    // fixed64 spanID = 2;
+    [LSBinaryCodec writeTagNumber:2 format:WireFormatFixed64 buffer:message];
+    [LSBinaryCodec writeFixedEncodedUInt64:spanID buffer:message];
+
+    // bool sampled = 3; (always true here)
+    [LSBinaryCodec writeTagNumber:3 format:WireFormatBool buffer:message];
+    [LSBinaryCodec writeVarintEncodedUInt64:YES buffer:message];
+
+    // map<string, string> baggage = 4;
+    [LSBinaryCodec writeStringMap:baggage tagNumber:4 buffer:message];
+
+    return message;
+}
+
++ (bool)decodeEmbeddedMessage:(NSData *)msg
+                         dest:(LSSpanContext *)dest
+                    withError:(NSError*__autoreleasing *)errorPtr {
+    UInt64 head = 0;
+    UInt64 tail = head + msg.length;
+    UInt64 traceId = 0;
+    UInt64 spanId = 0;
+    NSMutableDictionary *baggage = [NSMutableDictionary dictionary];
+
+    if (head >= tail) {
+        if (errorPtr) {
+            *errorPtr = [NSError errorWithDomain:OTErrorDomain code:OTInvalidCarrierCode userInfo:nil];
+        }
+        return false;
+    }
+
+    while (head < tail) {
```

Comment:
Yes, corrupt data could lead to an infinite loop here. There's also a bug where we weren't advancing the offset correctly after reading a baggage string.